### PR TITLE
[FLINK-12877][table][hive] Unify catalog database implementations

### PR DESCRIPTION
**Pull request by bowenli86**
## What is the purpose of the change

This PR unifies catalog database implementations.

## Brief change log

- unifies `GenericCatalogDatabase` and `HiveCatalogDatabase` into `CatalogDatabaseImpl`
- updates `HiveCatalog` accordingly
- removes `GenericCatalogDatabase`, `HiveCatalogDatabase`, and `AbstractCatalogDatabase`
- updates unit tests

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-table/flink-table-api-java/pom.xml [hunk 1]
```diff
@@ -58,4 +58,20 @@ under the License.
 			<artifactId>flink-test-utils-junit</artifactId>
 		</dependency>
 	</dependencies>
+
+	<build>
```
Comment (bowenli86) [inline flink-table/flink-table-api-java/pom.xml]: FYI, the indentation is correct. Seems to be github display problem

**Comment by bowenli86** [comment]
cc @xuefuz @lirui-apache @zjuwangg

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 2]
```diff
@@ -172,13 +179,23 @@ public void close() throws CatalogException {
...
+
+		// By default, isGeneric=false
+		boolean isGeneric = Boolean.valueOf(properties.remove(CatalogConfig.FLINK_IS_GENERIC));
+
+		if (isGeneric) {
+			properties = retrieveFlinkProperties(properties);
+		}
+
+		properties.put(HiveDatabaseConfig.DATABASE_LOCATION_URI, hiveDatabase.getLocationUri());
+		properties.put(CatalogConfig.FLINK_IS_GENERIC, String.valueOf(isGeneric));
```
Comment (lirui-apache) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java]: Why do we need to remove this property and then put it back?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 3]
```diff
@@ -228,32 +246,30 @@ public void alterDatabase(String databaseName, CatalogDatabase newDatabase, bool
...
+			if (!ignoreIfNotExists) {
+				throw new DatabaseNotExistException(getName(), databaseName);
+			}
 		} catch (TException e) {
 			throw new CatalogException(String.format("Failed to alter database %s", databaseName), e);
 		}
 	}
 
+	private static void validateAlterDatabase(Map<String, String> oldProp, Map<String, String> newProp) {
+		for (String immutable : IMMUTABLE_DATABASE_PROPERTIES) {
```
Comment (lirui-apache) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java]: Shall we also verify old DB and new DB are of same type?

**Comment by lirui-apache** [comment]
Thanks @bowenli86 for the PR. One question is do we plan to differentiate generic and hive-native databases? Seems HiveCatalog tries to tell them apart by checking `CatalogConfig.FLINK_IS_GENERIC`, but we don't have test to create generic DBs?

**Comment by bowenli86** [comment]
@lirui-apache thanks for your review! After a discussion with @xuefuz, we felt it's not necessary to distinguish whether a database is generic or not. Thus I completely removed that part from the PR.

@xuefuz can you take a look?

**Comment by bowenli86** [comment]
@xuefuz thanks for your review!

Merging

**Pull request closed without merging**